src/update_agent/actor.rs

```diff
@@ -308,9 +308,36 @@ impl UpdateAgent {
     ) -> ResponseActFuture<Self, Result<(), ()>> {
         trace!("trying to finalize an update");
 
-        let can_finalize = self.strategy.can_finalize();
-        let state_change = actix::fut::wrap_future::<_, Self>(can_finalize)
-            .then(|can_finalize, actor, _ctx| actor.finalize_deployment(can_finalize, release))
+        let strategy_can_finalize = self.strategy.can_finalize();
+        let state_change = actix::fut::wrap_future::<_, Self>(strategy_can_finalize)
+            .then(|strategy_can_finalize, actor, _ctx| {
+                if !strategy_can_finalize {
+                    // Reset number of postponements if strategy does not allow
+                    // finalization.
+                    update_unit_status(&format!(
+                        "update staged: {}; reboot pending due to update strategy",
+                        &release.version
+                    ));
+                    actor.state.update_staged(release);
+                    Box::pin(actix::fut::err(()))
+                } else {
+                    actor.check_interactive_sessions(release)
+                }
+            })
+            .then(|res, actor, _ctx| {
+                match res {
+                    // No interactive sessions detected.
+                    Ok(release) => {
+                        log::info!(
+                            "staged deployment '{}' available, proceeding to finalize it",
+                            release.version
+                        );
+                        actor.finalize_deployment(release)
+                    }
+                    // Interactive sessions detected or strategy does not permit finalization.
+                    Err(_) => Box::pin(actix::fut::err(())),
```

Comment:
This could have happened due to `!strategy_can_finalize` OR `check_interactive_sessions()` returned an error (there are interactive users and a reboot should be postoned).